### PR TITLE
UI Pointer Optimization

### PR DESCRIPTION
**Pull request by SamGondelman**
- Generalized Pick and Pointer classes, managers, and scripting interfaces
- Stylus Pick and Pointer C++ classes
- Pointers (lasers and styluses) automatically generate pointer events
- OffscreenQmlSurface touch event improvements
- Updated scripts to use new system
- Fixes scrolling on tablet with lasers ([FB8076](https://highfidelity.fogbugz.com/f/cases/8076))
- Fixes stylus interaction with the tablet ([FB8896](https://highfidelity.fogbugz.com/f/cases/8896))
- Improves tablet interaction as game rate drops ([FB7920](https://highfidelity.fogbugz.com/f/cases/7920/), [FB8392](https://highfidelity.fogbugz.com/f/cases/8392/))

Test plan:
- All grabbing tests: https://highfidelity.testrail.net/index.php?/suites/view/20
- All tablet interaction tests: https://highfidelity.testrail.net/index.php?/suites/view/27
- All context overlays tests: https://highfidelity.testrail.net/index.php?/suites/view/103
- Haptics should behave the same as in master.  Namely: strong pulse on click with lasers/stylus, small pulse when hovering over virtual keyboard letters.
- Create some web entities and use the hand lasers on them.  You should be able to scroll/click, etc.  
- If you trigger or press with the stylus on the tablet, you should be able to scroll off the surface.
- Interactions with lasers and stylus on the tablet should be faster and less glitchy.  Try in an empty domain and a heavily loaded one where your game rate is much lower (dev-welcome is good, or anywhere where your game rate is less than ~45 in HMD).
- Try at different avatar scales.  Stylus and lasers should still work on the tablet.
- Run all the LaserPointer tests in the hifi_tests repository: https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer/laser

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5015/HighFidelity-Beta-PR11834-ebe14fa6bb4279fe2fefab4fe78cca8feefcdafe.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5015/HighFidelity-Beta-PR11834-ebe14fa6bb4279fe2fefab4fe78cca8feefcdafe.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5025/HighFidelity-Beta-PR11834-30933d766696bc6b138339b185fbd9533e8dde3b.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5027/HighFidelity-Beta-PR11834-30933d766696bc6b138339b185fbd9533e8dde3b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5027/HighFidelity-Beta-PR11834-30933d766696bc6b138339b185fbd9533e8dde3b.dmg

**Comment by Nex-Pro** [comment]
Testing win10

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5041/HighFidelity-Beta-PR11834-61c7ccde6ed80b135edcdc6df672466ea4211c6d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5041/HighFidelity-Beta-PR11834-61c7ccde6ed80b135edcdc6df672466ea4211c6d.dmg

**Comment by YuppyPlays** [comment]
Testing.

**[Will start on single tester related tasks and be joined by Cain later later to finalize 2-3 person testers with our combined hardware]**

**Comment by YuppyPlays** [comment]
Testing notes (Will edit as I go along):

Went to start single-grab tests which involve Dev-content/testing domain, domain mismatch currently:
image

Checking other parts of the tests I can do in the meantime, will update this further shortly.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5067/HighFidelity-Beta-PR11834-033f658fff00f70702d06c81210dae4e3a01fbc1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5067/HighFidelity-Beta-PR11834-033f658fff00f70702d06c81210dae4e3a01fbc1.dmg

**Comment by YuppyPlays** [comment]
Testing complete.

Please see my previous comment for full information, please let me know if there is more I can do/provide. :) This was a long but rewarding, mainly passes, few fails, but I feel that means unfortunately this is a **Fail** til the issues are looked into.

Thank you @SamGondelman for the very quick assistance regarding this PR, :D 

Tested w/ Cain for multi-tester segments.

**Comment by SamGondelman** [comment]
@YuppyPlays thank you so much!  I've read through all your test notes.  It looks like the issues are:
- glitchiness with near grab
- avatar store grab issues
- the create app bringing up the keyboard erroneously
- the tablet sizing issue (non-blocking, and I think that should be fixed here, actually: https://github.com/highfidelity/hifi/pull/11847)
- laser pointer tests not running

I'll ask @druiz17 to look at the first three since they sound grab and qml related (and might also be present in master, too).  For the laser pointer tests, do you see anything in your logs when you try to run them?

Glad to hear that, in general, the lasers and stylus seemed to work!

**Comment by divyaelango88** [comment]
Thanks for taking a look @SamGondelman . There is also one additional failure with Create - https://highfidelity.testrail.net/index.php?/tests/view/14602&group_by=cases:section_id&group_order=asc&group_id=28

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5092/HighFidelity-Beta-PR11834-76935c145721ad841a0d4e5545b8fe21e1f94978.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5092/HighFidelity-Beta-PR11834-76935c145721ad841a0d4e5545b8fe21e1f94978.dmg

**Comment by jherico** [comment]
@YuppyPlays when you say you tried to load the test via URL, do you mean you put `https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer/laser` into the URL field?  That absolutely wont work...  that's just a folder with a bunch of subfolders and files.  You would need to clone the repository or download it as a zip file and load the test files from the local filesystem.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5095/HighFidelity-Beta-PR11834-a201312dcfc82e6b739b4ad8403553d2efef5877.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5095/HighFidelity-Beta-PR11834-a201312dcfc82e6b739b4ad8403553d2efef5877.dmg

**Comment by divyaelango88** [comment]
@YuppyPlays & @cain-kilgore the issues reported with Grab and the Create button popping up a virtual keyboard happen on master already. If you can quickly run those scripts per Austin's instruction in the comment above we should have this QA approved.

File: libraries/script-engine/src/ScriptCache.cpp [hunk 1]
```diff
@@ -69,6 +69,7 @@ void ScriptCache::getScriptContents(const QString& scriptOrURL, contentAvailable
     #ifdef THREAD_DEBUGGING
     qCDebug(scriptengine) << "ScriptCache::getScriptContents() on thread [" << QThread::currentThread() << "] expected thread [" << thread() << "]";
     #endif
+    forceDownload = true;
```
Comment (samcake) [inline libraries/script-engine/src/ScriptCache.cpp]: why is that for ?

Comment (samcake) [reply libraries/script-engine/src/ScriptCache.cpp]: is it expected that we force it to download ?

Comment (druiz17) [reply libraries/script-engine/src/ScriptCache.cpp]: I do not have a clue.

**Comment by cain-kilgore** [comment]
# Test Passed
Tested lasers, looks good. Confirmed the tablet problems mentioned earlier by @YuppyPlays but not related to this PR it seems.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5103/HighFidelity-Beta-PR11834-7a5dc9e00b3440dd533efd5d678734f71fa4c941.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5103/HighFidelity-Beta-PR11834-7a5dc9e00b3440dd533efd5d678734f71fa4c941.dmg

**Comment by YuppyPlays** [comment]
@SamGondelman 
No problem at all, thanks for having me :P haha

I think you're on point with the notes, regarding the lasers I believe Cain above has tested this as I was away ( @cain-kilgore - Much appreciated by the way, I was tied up yesterday with some other stuff away from my desk)

@jherico 
Ah yes I did do it that way, apologies, it was not specified I would need to do this as far as I recollect, though this is done now I appreciate you letting me know :)

@divyaelango88 
Ah I see, no worries, Cain seems to have tested for this.


Apologies for such a delay, have been away from my desk, if I'm required to re-test please do let me know.

**Comment by hyperlogic** [comment]
Looks good, thanks for creating the jsdoc comments, as well. 👍

**Comment by divyaelango88** [comment]
we need to do one more round of testing with this updated Test plan. @YuppyPlays or @cain-kilgore can you help please https://highfidelity.testrail.net/index.php?/suites/view/20&group_by=cases:section_id&group_order=asc&group_id=13

**Comment by YuppyPlays** [comment]
Testing complete:

Tested w/ Cain, unfortunately a **Fail**, TestRail Results: https://highfidelity.testrail.net/index.php?/runs/view/1148&group_by=cases:section_id&group_order=asc

**Comment by samcake** [comment]
waiting for this to build, test the build myself and merge if  it is correct

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5116/HighFidelity-Beta-PR11834-a0e40efdc85ac01a62ea4b0515f06c90a8a27a8f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5116/HighFidelity-Beta-PR11834-a0e40efdc85ac01a62ea4b0515f06c90a8a27a8f.dmg

**Pull request merged**